Improve CSV File Adapter error handling

danotorrey opened the pull request:
## Description
<!--- Describe your changes in detail -->
Fix the following issues in the CSV File Lookup Adapter:
1) Previously, a NullPointerException would happen the the CSV adapter file was not writable (moved, or otherwise inaccessible), and then writable once again on a subsequent refresh cycle. The `fileInfo` would become disconnected due to the file change (and refresh while in that state), and the next refresh would trigger a NPE, since the `fileInfo` would then be null.
2) Blank lines within the CSV (especially the common case of a blank line at the end of the file) would trigger a `Index Out of Bounds` error when attempting to read the value column (since it does not exist).  Blank lines are now gracefully skipped. 
3) When line exist in the CSV file that don't have the needed key and value columns would also trigger an `Index Out of Bounds` error. Extra handling has been added to catch this condition and return a valid/detailed error to the user in the log and UI. This is considered a fatal error, which must be corrected by the user, so the adapter will enter an error state when this is detected. 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Closes https://github.com/Graylog2/graylog-plugin-enterprise/issues/3821

While fixing the above issue, I found a few other issues, that I fixed at the same time.
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Manual testing, some unit test coverage.

## Testing
Set a low refresh interval (e.g. 10 seconds). Set the adapter up correctly and verify it works. Move the CSV file to a different location. Let the adapter go through one refresh cycle so that it logs the `The specified file [%s] does not exist or is not writable` error. Restore access to the file and verify that the adapter recovers successfully.

Verify that blank value lines (especially at the end of file) are now ignored.

Add some invalid lines without valid key and value columns. Verify those conditions are handled.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

danotorrey commented on graylog2-server/src/main/java/org/graylog2/lookup/adapters/CSVFileDataAdapter.java:
```diff
@@ -175,22 +189,43 @@ private Map<String, String> parseCSVFile() throws IOException {
             LOG.error("Couldn't parse CSV file {} (settings separator=<{}> quotechar=<{}> key_column=<{}> value_column=<{}>)", config.path(),
                     config.separator(), config.quotechar(), config.keyColumn(), config.valueColumn(), e);
             setError(e);
+            throw new IllegalStateException(e);
```
Parsing errors were being silently ignored in the previously, because `clearError` on line 147 was clearing it out. Throwing the exception here properly reports the unrecoverable error to the UI:

![image](https://user-images.githubusercontent.com/3423655/218857102-09c934ef-552c-405f-b44f-dfb88d88bc5c.png)

ryan-carroll-graylog reviewed (approved):
Looks good and tests successfully!

kingzacko1 reviewed (commented):
Looks good and tests successfully. There is identical code in `IlluminateCSVFileDataAdapter` that could also likely use these changes, even though it is unlikely we ship a faulty CSV with Illuminate, it isn't impossible.

danotorrey commented:
Thanks @kingzacko1! Good point. I will check that adapter too, and add these improvements if needed.

danotorrey commented:
@kingzacko1 I have added the same ability to skip blank lines to the Illuminate CSV adapter in https://github.com/Graylog2/graylog-plugin-enterprise/pull/4745. This also jogged my memory that there is a Cloud Upload CSV file adapter too, so I added the same improvements there too: https://github.com/Graylog2/graylog-plugin-cloud/pull/1092.

danotorrey commented:
@ryan-carroll-graylog @kingzacko1 Just wanted to give a heads up, since I am about to request reviews on a bunch of PRs related to this. I summarized them here for clarity.

1) I have filed two small fix PRs for an import issue that found while working on backports: 
https://github.com/Graylog2/graylog2-server/pull/14781
https://github.com/Graylog2/graylog-plugin-enterprise/pull/4760

2) Backport PRs were filed for the CSV, Illuminate CSV, and Cloud CSV adapter backports (both 5.0 and 4.3 since it's a bug fix). Note that Cloud only has a 5.0 backport, since the 4.3 branch seems out-of-date. I think Cloud deploys only happen with 5.0 now. 
**Standard CSV 5.0/4.3**
https://github.com/Graylog2/graylog2-server/pull/14779
https://github.com/Graylog2/graylog2-server/pull/14780
**Illuminate CSV 5.0/4.3**
https://github.com/Graylog2/graylog-plugin-enterprise/pull/4758
https://github.com/Graylog2/graylog-plugin-enterprise/pull/4759
**Cloud upload 5.0**
https://github.com/Graylog2/graylog-plugin-cloud/pull/1094
(no 4.3 for cloud due to unmaintained 4.3 branch likely not being used)
